### PR TITLE
Typed list implement and expose allocation

### PR DESCRIPTION
**Pull request by esc**
This implements pre-allocating a typed-list and checking the allocation from all levels. Tests included.

### CODE REVIEW

File: numba/tests/test_listobject.py [hunk 1]
```diff
@@ -37,6 +37,25 @@ def foo(n):
...
 
+class TestAllocation(MemoryLeakMixin, TestCase):
+
+    def test_list_allocation(self):
+        @njit
+        def foo_kwarg(n):
+            l = listobject.new_list(int32, allocated=n)
+            return l._allocated()
+
+        for i in range(16):
```
Comment (stuartarchibald) [inline numba/tests/test_listobject.py]: what about if the allocation is specified as negative?

File: numba/tests/test_typedlist.py [hunk 2]
```diff
@@ -445,6 +445,40 @@ def setup(start=10, stop=20):
             self.assertEqual(rl, list(tl))
 
 
+class TestAllocation(MemoryLeakMixin, TestCase):
+
+    def test_allocation(self):
+        # kwarg version
+        for i in range(16):
+            tl = List.empty_list(types.int32, allocated=i)
+            self.assertEqual(tl._allocated(), i)
```
Comment (stuartarchibald) [inline numba/tests/test_typedlist.py]: -ve allocation value?

File: numba/cext/listobject.c [hunk 3]
```diff
@@ -45,6 +45,7 @@
  * The following additional functions are implemented for the list, these are
  * needed to make the list work within Numba.
  *
+ * - Accessing the allocation numba_list_allocated
```
[the remark this answers is not in the document]
Comment (esc) [reply numba/cext/listobject.c]: Indeed this exposes the number of pre-allocated slots that are allocated for future appends to the list.

**Comment by esc** [comment]
@stuartarchibald I fixed the issues you mentioned. As discussed in private communication, we will keep `allocated` as this is what cpython uses originally: 

https://github.com/python/cpython/blob/master/Include/listobject.h#L39

File: numba/tests/test_listobject.py (hunk 1 above)
Comment (esc) [reply numba/tests/test_listobject.py]: has been fixed and test added.

File: numba/tests/test_typedlist.py (hunk 2 above)
Comment (esc) [reply numba/tests/test_typedlist.py]: has been fixed and test added.

**Comment by esc** [comment]
@stuartarchibald I have addressed all the comments and this is ready for the next round of inspection.

**Pull request merged**